### PR TITLE
Added `TimeoutOptions` to `SwarmOptions`

### LINKED ISSUE #1957: Group timeout related options
See #1933 and its comments.

Seems like a good starting point. For the most part, as far as I know, even [9c-headless] doesn't use variable arguments for timeout options. 🙄

[9c-headless]: https://github.com/planetarium/NineChronicles.Headless

### PR DESCRIPTION
**Pull request by greymistcube**
Closes #1957.

Had to update certain parts of `Swarm<T>` API to make it less non-sensical. 🙄
There's still a long way to go to mitigate #1933.

### CODE REVIEW

**Comment by longfin** [comment]
Well... `TimeoutOptions` class seemed that classify its format. IMHO, many of classification is more natural for me when classify by semantics. for example, Containing `.GetBlockHashesTimeout` and `.GetBlocksBaseTimeout` in the same class is natural for me, but `.BootstrapDialTimeout` and `.GetBlockHashesTimeout` isn't... 😕 

Is there any useful case for grouping all timeouts in one class? (Ignoring the usage and wanting to manipulate only the whole timeout)

**Comment by greymistcube** [comment]
@longfin
I think it depends on how we interpret it. Yes, `BootstrapDialTimeout` is kind of different from `GetBlockHashesTimeout`, but these are ultimately all used **only** to be passed on to `ITransport.SendMessageAsync()`, and in that sense they are also kind of the same. As there are three phases for `Swarm<T>`, namely `BootstrapAsync()`, `PreloadAsync()`, and `StartAsync()`, and each phase has different set of `Message` types that they deal with, if we take this to the extreme, we might get something like:

- `SwarmOptions`
  - `TimeoutOptions`
    - `BootstrapOptions`
      - `PingTimeout`
      - `FindNeighborsTimeout`
    - `PreloadOptions`
      - `PingTimeout`
      - ...

You get the idea.

Another way to organize this would be:

- `SwarmOptions`
  - `BootstrapOptions`
    - `TimeoutOptions`
      - `PingTimeout`
      - `FindNeighborsTimeout`
    - `PreloadOptions`
     - `PingTimeout`
     - ...

In this case, yes, it kind of feels unnatural to have `BootstrapDialTimeout` in the same hierarchy as `GetBlockHashesTimeout`. I do agree the latter design would probably be better further down the road, but I decided to go with the former for the following reasons:

- I was lazy; I didn't want to create more than two layers at this point.
- It seemed silly to have `BootstrapDialTimeout` in `SwarmOptions`.
- Again, I was lazy; separating out `SwarmOptions`, `BootstrapOptions`, `PreloadOptions`, `AsyncOptions` seemed like too much work. Couldn't figure out how there'd be a nice sensible workable unit for an iterative process.

> Is there any useful case for grouping all timeouts in one class? (Ignoring the usage and wanting to manipulate only the whole timeout)

Not really. 😗
Joking aside, from usage perspective, for the most part, I thought it'd be better to just **dump everything** during `Swarm<T>()` and make it just work if possible. I just had to stick it in **somewhere** so I could work towards `Swarm<T>.BootstrapAsync()`, `Swarm<T>.PreloadAsync()`, etc. without any arguments.

**Comment by longfin** [comment]
> I was lazy; I didn't want to create more than two layers at this point.

Agree, I also think that we need to flatten hierarchy as possible. so more comportable structure in my mind was...

- SwarmOptions
  - BootstrapOptions
    - PingTimeout
    - FindNeighborsTimeout
  - PreloadOptions
    - PingTimeout
  - BlockSyncOptions(?)
    - GetBlockHashesTimeout 

But... well, OK I realize that, there is too many `dialTimeout`, `pingTimeout` and any other timeouts in [`Swarm<T>` API](https://docs.libplanet.io/0.34.0/api/Libplanet.Net.Swarm-1.html). In this situation, I think it might not be strange to tie them together in terms of waiting for responses from other peers. so I also think it might be a good idea to group them by request/response message type. (I think the options are organized in this PR with this perspective)

However on the other side, I still haven't found a reason to separate these timeouts by layering them in `SwarmOption` (`Options.TimeoutOptions.DialTimeout` instead of `Options.DialTimeout`), other than a visual cleanup of the code...

**Comment by greymistcube** [comment]
> However on the other side, I still haven't found a reason to separate these timeouts by layering them in SwarmOption (Options.TimeoutOptions.DialTimeout instead of Options.DialTimeout), other than a visual cleanup of the code...

It's just a starting point to have at least some semblance of organization. Neither am I particularly happy with the current result at the moment. 😐

I wanted to track down each option and:

- Figure out what it does and what it is meant to do as to clarify its usage
- Fix internal inconsistency, inappropriate uses, and/or remove redundancy

However, doing this to a whole of `SwarmOptions` turns out to be lot more tedious and time consuming than it seems. I just needed to have some "scope" to work with, and put them somewhere. This probably will be temporary and ultimately `TimeoutOptions` might be short lived and I also agree there is a bit of code smell, I really don't think it stinks any more than the current state `SwarmOptions` is in; in all honesty, I consider it to be a complete mess anyway in its current form.

On the other hand, I don't quite agree with keeping `SwarmOptions` completely flat. 😕

**Review by longfin — DISMISSED** [review]

**Review by riemannulus — APPROVED** [review]

File: CHANGES.md [hunk 1]
```diff
@@ -48,21 +69,19 @@ To be released.
 
  -  The following store URI schemes are deprecated:  [[#1573], [#1955]]
 
-     - `default`: Use `default+file` instead.
-     - `rocksdb`: Use `rocksdb+file` instead.
-
```
Comment (dahlia) [inline CHANGES.md:53]: Why were these lines removed?

File: CHANGES.md [hunk 2]
```diff
@@ -48,21 +69,19 @@ To be released.
...
+[#1957]: https://github.com/planetarium/libplanet/issues/1957
 [#1959]: https://github.com/planetarium/libplanet/issues/1959
 [#1960]: https://github.com/planetarium/libplanet/pull/1960
 [#1961]: https://github.com/planetarium/libplanet/pull/1961
+[#1962]: https://github.com/planetarium/libplanet/pull/1962
 [#1966]: https://github.com/planetarium/libplanet/pull/1966
 
 
 Version 0.34.0
---------------
```
Comment (dahlia) [inline CHANGES.md:65]: You probably made some mistakes during rebasing.

**Review by dahlia — CHANGES_REQUESTED** [review]

File: CHANGES.md (hunk 1 above)
Comment (greymistcube) [reply CHANGES.md:53]: I'm fixing it currently. I accidently removed some links also while rebasing. 😅

File: CHANGES.md [hunk 3]
```diff
@@ -27,6 +44,10 @@ To be released.
  -  Added `NameValueCollectionExtensions` static class.  [[#1955]]
  -  Added `TimeSpanExtensions.Multiply()` static method.  [[#1966]]
  -  Type support for `Guid` added to `DataModel`.  [[#1959], [#1960]]
+ -  `TimeSpanExtensions.Multiply()` method added.  [[#1966]]
```
Comment (greymistcube) [inline CHANGES.md:47]: Still needs a fix. 😑

File: CHANGES.md (hunk 2 above)
Comment (greymistcube) [reply CHANGES.md]: Should be all fixed I think...? Can you check again for me? 😅

**Review by dahlia — APPROVED** [review]
